### ISSUE TITLE
The FFI-chapter needs some clarification on addressof and C

### ISSUE THREAD

**Issue by marsuk** [ponylang/pony-tutorial#332]
In Chapter 9 "C FFI" [](https://tutorial.ponylang.io/c-ffi/calling-c.html) in the paragraph "C types", after saying "just like taking an address in C" it might be good to say something like this:

Note that in C, when you write only the variable name of a struct instance, C's address operator & gives you the pointer to the first member. This is not the case in Pony. You have to state the first member of the struct, otherwise it results in a pointer that points to the pointer to the first member.

**Comment by jemc** [comment]
I don't really follow the particular example text you gave.

Can you elaborate on what you think is missing from the original description?

**Comment by marsuk** [comment]
It's just that Pony doesn't do what you would expect when you have used C before.

In C: `&struct_instance` == `&struct_instance.first_member`
But in Pony: `adressof struct_instance` != `adressof struct_instance.first_member`

This means, in FFI Pony code you currently cannot just write `addressof window_class` or anything like that, because it won't work. I had to write `addressof window_class.cbSize` to get the correct address of the structure. `cbSize` is the first field in the structure.

The address that you get with `addressof window_class` is something unrelated, maybe something that Pony uses internally. It happens to be another pointer pointing to the what I actually wanted. Maybe mentioning this in the tutorial would be too confusing, or this ponyc behaviour might change anyway.

When the behaviour of Pony's `addressof` is what you want it to be and is not a bug, for whatever reason, then it would be good to say in the tutorial that Pony users need to be aware of it and that it will not look like C code looks like.

**Comment by SeanTAllen** [comment]
Can you supply a small snippet code that can be used to see what you are talking about @marsuk ?

**Comment by marsuk** [comment]
Yes, in the issue https://github.com/ponylang/ponyc/issues/2944 I was asking how I would eventually create a Windows application, and there is code like this. Although, that issue was not about this problem here.

That's the shorter one of the two versions in the issue:
https://gist.github.com/ponylang-gist/4ee25805123c8c646ce5fc6a1db93975

When you change line 97 from
`let class_atom: ATOM = @RegisterClassExA[ATOM](addressof wndclass.cbSize)`
to
`let class_atom: ATOM = @RegisterClassExA[ATOM](addressof wndclass) // .cbSize)`
the application will not start anymore.

**Comment by ergl** [comment]
@SeanTAllen I think this can be closed since it was partially fixed by #474, which added the following:

```
Pony classes and structs correspond directly to pointers to the class or struct in C.

...

To pass pointers to values to C the `addressof` operator can be used
(previously `&`), just like taking an address in C. This is done in the
standard library to pass the address of a `U32` to an FFI function that
takes a `int*` as an out parameter:

...

### Working with Structs: from Pony to C

Like we mentioned above, Pony classes and structs correspond directly
to pointers to the class or struct in C. This means that in most cases we
won't need to use the `addressof` operator when passing struct types to C.
For example, let's imagine we want to use the `writev` function from Pony on Linux:

...

```
